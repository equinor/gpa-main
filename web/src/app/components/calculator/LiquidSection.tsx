import styled from 'styled-components/macro';
import React from 'react';
import { FormSection } from '../ui/FormSection';
import { Label, NumberInput } from '../ui/Inputs';


export interface Liquid {
  nitrogen: number,
  methane: number,
  ethane: number,
  propane: number,
  iButane: number,
  nButane: number,
  iPentane: number,
  nPentane: number,
  nHexane: number,
}

interface LiquidSectionProps {
  liquid: Liquid,

  setLiquid(liquid: Liquid): void,
}

const LiquidInputs = styled.span`
  display: flex;
  flex-direction: row;
`;

const LiquidInput = styled.p`
  margin-right: 25px;
`;

interface LiquidComponentInputProps {
  componentName: string,
  componentValue: number,

  onChange(value: number): void,
}

const LiquidComponentInput: React.FC<LiquidComponentInputProps> = ({ onChange, componentName, componentValue }) => {
  return (
    <LiquidInput>
      <Label htmlFor={`liquid-${componentName}`}>
        {componentName}
      </Label>
      <NumberInput
        id={`liquid-${componentName}`}
        value={componentValue.toFixed(2)}
        onChange={e => onChange(parseFloat(e.target.value))}
        step='0.01'
      />
    </LiquidInput>
  );
};

export const LiquidSection: React.FC<LiquidSectionProps> = ({ liquid, setLiquid }) => (
  <FormSection legendText='Liquid' index={2}>
    <LiquidInputs>
      <div>
        <h3>Standard gases</h3>
        <LiquidComponentInput
          componentName='nitrogen'
          componentValue={liquid.nitrogen}
          onChange={value => setLiquid({ ...liquid, nitrogen: value })}
        />
      </div>
      <div style={{ flexGrow: 1, height: '180px' }}>
        <h3>Light hydrocarbons</h3>
        <div style={{ display: 'flex', flexDirection: 'column', flexGrow: 1, flexWrap: 'wrap', height: '100%', justifyContent: 'flex-start', alignContent: 'flex-start' }}>
          {
            (Object.keys(liquid) as Array<keyof typeof liquid>).map((componentName, index) => componentName !== 'nitrogen' && (
              <LiquidComponentInput
                key={index}
                componentName={componentName}
                componentValue={liquid[componentName]}
                onChange={value => setLiquid({ ...liquid, [componentName]: value })}
              />
            ))
          }
        </div>
      </div>
    </LiquidInputs>
  </FormSection>
);