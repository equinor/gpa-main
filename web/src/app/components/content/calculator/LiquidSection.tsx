import styled from 'styled-components/macro';
import React from 'react';
import { FormSection } from '../../ui/FormSection';
import { H3 } from '../../elements/Texts';
import { StandardInput } from '../../elements/Inputs';


export interface ILiquid {
  nitrogen: number,
  methane: number,
  ethane: number,
  propane: number,
  iButane: number,
  nButane: number,
  iPentane: number,
  nPentane: number,
  nHexane: number,
}

interface ILiquidSectionProps {
  liquid: ILiquid,

  setLiquid(liquid: ILiquid): void,
}

export const LiquidSection: React.FC<ILiquidSectionProps> = (props) => {
  return (
    <FormSection legendText='Liquid' index={2}>
      <StLiquidInputs>
        <div>
          <H3>Standard gases</H3>
          <LiquidComponentInput
            componentName='nitrogen'
            componentValue={props.liquid.nitrogen}
            onChange={(value: any) => {
              setLiquid("nitrogen", value);
            }}
            onBlur={(value: any) => {
              setLiquid("nitrogen", value, true);
            }}
            required={false}
          />
        </div>
        <div style={{ flexGrow: 1, height: "100%" }}>
          <H3>Light hydrocarbons</H3>
          <StMatrixContainer>
            {
              (Object.keys(props.liquid) as Array<keyof typeof props.liquid>).map((componentName, index) => componentName !== 'nitrogen' && (
                <LiquidComponentInput
                  key={index}
                  componentName={componentName}
                  componentValue={props.liquid[componentName]}
                  onChange={(value: any) => {
                    setLiquid(componentName, value);
                  }}
                  onBlur={(value: any) => {
                    setLiquid(componentName, value, true);
                  }}
                  required={false}
                />
              ))
            }
          </StMatrixContainer>
        </div>
      </StLiquidInputs>
    </FormSection>
  )

  function setLiquid(name: keyof ILiquid, value: any, toFixed?: boolean) {
    let liquid: ILiquid = props.liquid;
    if (value !== "") {
      const formattedValue = toFixed ? parseFloat(parseFloat(value).toFixed(2)) : parseFloat(value);
      liquid[name] = formattedValue as never;
      props.setLiquid({ ...props.liquid });
    }
    else {
      liquid[name] = null as never;
      props.setLiquid({ ...props.liquid });
    }
  }
}

interface ILiquidComponentInputProps {
  componentName: string;
  componentValue: number;

  onChange(value: number): void;
  onBlur(value: number): void;
  required: boolean;
}

const LiquidComponentInput: React.FC<ILiquidComponentInputProps> = (props) => {
  return (
    <StLiquidInput style={
      props.componentName === "propane" || props.componentName === "iPentane" ? {marginBottom: "0px"} : {}
    }>
      <StandardInput
        id={`liquid-${props.componentName}`}
        label={props.componentName}
        placeholder={props.componentName}
        value={props.componentValue}
        type="number"
        onChange={(e: any) => {
          props.onChange(e.target.value)
        }}
        onBlur={(e: any) => {
          props.onBlur(e.target.value)
        }}
        required={props.required}
      ></StandardInput>
    </StLiquidInput>
  );
};

const StLiquidInputs = styled.span`
  display: flex;
  flex-direction: row;
  h3 {
    margin: 10px 0 15px 0;
  }
`;

const StLiquidInput = styled.div`
      margin: 0 30px 15px 0;
      width: 135px;
`;

const StMatrixContainer = styled.div`
  display: flex;
  flex-grow: 1;
  flex-wrap: wrap; 
  height: 220px; 
  width: 470px;
  justify-content: flex-start;
  flex-direction: column;
`;