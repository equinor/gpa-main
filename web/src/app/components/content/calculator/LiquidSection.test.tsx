import { MockedProvider } from '@apollo/react-testing';
import React from 'react';
import renderer from 'react-test-renderer';

import { ILiquid, LiquidSection } from './LiquidSection';

describe("LiquidSection", () => {
    //loading
    it("snapshots", () => {
        const liquid: ILiquid = {
            nitrogen: 5,
            methane: 5,
            ethane: 5,
            propane: 5,
            iButane: 5,
            nButane: 5,
            iPentane: 5,
            nPentane: 5,
            nHexane: 5,
        };
        const tree = renderer.create(
            <MockedProvider addTypename={false}>
                <LiquidSection
                    liquid={liquid}
                    setLiquid={() => { }}
                />
            </MockedProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })
}); 