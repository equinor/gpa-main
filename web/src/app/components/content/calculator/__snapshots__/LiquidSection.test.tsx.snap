// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiquidSection snapshots 1`] = `
<div
  className="FormSection__StFormSection-sc-4jzlq-0 ccofqN"
>
  <span
    className="FormSection__StCircleIcon-sc-4jzlq-2 fppCCB"
  >
    <span>
      2
    </span>
  </span>
  <fieldset
    className="FormSection__StFieldSet-sc-4jzlq-3 flTkdv"
  >
    <legend
      className="FormSection__StLegend-sc-4jzlq-1 fvKqdS"
    >
      Liquid
    </legend>
    <span
      className="LiquidSection__StLiquidInputs-x68rdr-0 dcIfbi"
    >
      <div>
        <h3
          className="Texts__H3-teybz3-1 iRQgzB"
        >
          Standard gases
        </h3>
        <div
          className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
          style={Object {}}
        >
          <div
            style={
              Object {
                "display": "flex",
                "flexWrap": "wrap",
                "justifyContent": "space-between",
              }
            }
          >
            <label
              className="Inputs__StLabel-dfit48-1 fWudmg"
              htmlFor="liquid-nitrogen"
            >
              <span>
                nitrogen
              </span>
            </label>
            <input
              className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
              id="liquid-nitrogen"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="nitrogen"
              required={false}
              step="0.01"
              type="number"
              value={5}
            />
          </div>
        </div>
      </div>
      <div
        style={
          Object {
            "flexGrow": 1,
            "height": "100%",
          }
        }
      >
        <h3
          className="Texts__H3-teybz3-1 iRQgzB"
        >
          Light hydrocarbons
        </h3>
        <div
          className="LiquidSection__StMatrixContainer-x68rdr-2 cHibXw"
        >
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={Object {}}
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-methane"
              >
                <span>
                  methane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-methane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="methane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={Object {}}
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-ethane"
              >
                <span>
                  ethane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-ethane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="ethane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={
              Object {
                "marginBottom": "0px",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-propane"
              >
                <span>
                  propane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-propane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="propane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={Object {}}
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-iButane"
              >
                <span>
                  iButane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-iButane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="iButane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={Object {}}
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-nButane"
              >
                <span>
                  nButane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-nButane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="nButane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={
              Object {
                "marginBottom": "0px",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-iPentane"
              >
                <span>
                  iPentane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-iPentane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="iPentane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={Object {}}
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-nPentane"
              >
                <span>
                  nPentane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-nPentane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="nPentane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
          <div
            className="LiquidSection__StLiquidInput-x68rdr-1 cTiNfU"
            style={Object {}}
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                }
              }
            >
              <label
                className="Inputs__StLabel-dfit48-1 fWudmg"
                htmlFor="liquid-nHexane"
              >
                <span>
                  nHexane
                </span>
              </label>
              <input
                className="Inputs__StInput-dfit48-2 Inputs__StNumberInput-dfit48-5 BTmgF"
                id="liquid-nHexane"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="nHexane"
                required={false}
                step="0.01"
                type="number"
                value={5}
              />
            </div>
          </div>
        </div>
      </div>
    </span>
  </fieldset>
</div>
`;
